src/FileSig.hs

```diff
@@ -7,126 +7,143 @@
   , prettyMatchPrint
   )  where
 
-import Data.Foldable
-import Data.FileEmbed
 import Paths_filesig
-import Data.Vector as V
-import Data.Maybe
-import Data.Functor
-import Data.Traversable
-import Control.Monad
-import Data.Text as T
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import Data.ByteString.Base16 (decodeBase16Lenient)
 import Data.HashMap.Strict as HM
-import Data.Aeson
-import Data.Aeson.Types
-import Control.Monad.IO.Class
-import Data.List                            as L
-import Data.Ord
-import qualified Data.Hex                   as H
-import qualified Data.ByteString            as BS
-import qualified Data.ByteString.Char8      as B8
-import qualified Data.ByteString.Lazy as B
-
--- | Magic bytes/file signature, in the form of the hex (`String` [no spaces]), followed
--- by the offset (`Int`).
-type Signature = (String, Int)
+import Data.List as L (sort)
+import Data.Ord (Down(..))
+import Data.Text as T (Text, pack, unlines, unpack)
+import System.IO (IOMode(..), hClose, withBinaryFile)
+
+
+import Magic
+
+
+-- | Magic bytes/file signature, in the form of the offset (`Int`), followed
+-- by the bytes (`ByteString`).
+type Signature = (Int, ByteString)
 
 -- | The extension of a file format, which acts as a key in the magic database.
 type ExtensionKey = Text
 
+-- | The MIME type/media type
+type MIME = Text
+
 -- | File format located by a file extension (as key) in the magic number database (JSON).
 data FileFormat = FileFormat {
   magicBytes :: [Signature],
   -- ^ The magic numbers/file signatures which can be used to identify this file format.
-  mime :: Text
+  mime :: MIME
   -- ^ the MIME type/media type of this file format.
 } deriving (Show)
 
 -- | The abstraction of the JSON magic database containing all the file signatures.
 -- The key is the file extension.
 type MagicMap = HM.HashMap ExtensionKey FileFormat
 
--- | JSON parser for a file format entries in the magic database.
-instance FromJSON FileFormat where
-  parseJSON = withObject "FileFormat" $ \o ->
-    FileFormat <$> fmap parseSignatures (o .: "signs")
-            <*> o .: "mime"
+-- | This is the entire MagicMap to be used everywhere, no IO required.
+--
+-- The 'decodeBase16Lenient' interprets the hexidecimal into raw bytestring,
+-- So it is already in optimal form at runtime.
+magicMap :: MagicMap
+magicMap =
+    HM.map go rawMagicMap
+  where
+    go :: ([Signature], Text) -> FileFormat
+    go (sigs, mime) = FileFormat (fmap decodeBase16Lenient <$> sigs) mime
+
+-- | All extensions in the 'magicMap'
+allExtensions :: [ExtensionKey]
+allExtensions = L.sort $ HM.keys magicMap
+
+-- This is going to be '(36865, "4344303031")' from the 'iso' signature
+-- but if the magicMap is ever updated, this will consistently give the
+-- biggest starting chunk we might need to guess extentions, so as to not
+-- load in potential GBs of data into memory.
+longestSignature :: Int
```

Comment:
Why are we getting the longest signature and not just sorting the signatures by length? I don't see the use/application here. The comments before and earlier commit history on sorting by signature length was in case two signatures matched in the beginning but the longer one differed at the end. Just like, for example, say a string is `haskell` and you're looking to match that in some lookup; how you would sort `has` vs. `haskell`; if you didn't have it sort by the longest first it would match `has` before it got a chance to match `haskell`.

Still drinking my morning coffee, but I hope that still made sense.

Reply:
`longestSignature` is used here to figure out what the largest chunk is that could possibly be needed to verify a file, so that when reading in the firstN bytes of a file, you'd only get what is needed in the worst case.
The longest one in this case is the `"iso"` signature, which might look for magic numbers some 38kb into the file, but that's the largest chunk out of all of them. So if someone wants to verify a 10GB+ movie/iso file, we just load the `longestSignature` amount of bytes into memory, so as to not have to load in the entire 10GB+ file (which would not be appreciated by users, I think)

Reply:
Wow that's a really good optimization!

Are the signatures already sorted by length, in order to address my previous comment (so a shorter signature isn't matched before a longer one has a chance)?